scene/resources/curve.cpp

```diff
@@ -1247,10 +1277,59 @@ void Curve3D::_bake() const {
 	baked_tilt_cache.resize(pointlist.size());
 	PoolRealArray::Write wt = baked_tilt_cache.write();
 
+	baked_up_vector_cache.resize(up_vector_enabled ? pointlist.size() : 0);
+	PoolVector3Array::Write up_write = baked_up_vector_cache.write();
+
+	Vector3 sideways;
+	Vector3 up;
+	Vector3 forward;
+
+	Vector3 prev_sideways = Vector3(1, 0, 0);
+	Vector3 prev_up = Vector3(0, 1, 0);
+	Vector3 prev_forward = Vector3(0, 0, 1);
+
 	for (List<Plane>::Element *E = pointlist.front(); E; E = E->next()) {
 
 		w[idx] = E->get().normal;
 		wt[idx] = E->get().d;
+
+		if (!up_vector_enabled) {
+			idx++;
+			continue;
+		}
+
+		forward = idx > 0 ? (w[idx] - w[idx - 1]).normalized() : prev_forward;
+
+		float y_dot = prev_up.dot(forward);
+
+		if (y_dot > (1.0f - CMP_EPSILON)) {
+			sideways = prev_sideways;
+			up = -prev_forward;
+		} else if (y_dot < -(1.0f - CMP_EPSILON)) {
+			sideways = prev_sideways;
+			up = prev_forward;
+		} else {
+			sideways = prev_up.cross(forward).normalized();
+			up = forward.cross(sideways).normalized();
+		}
+
+		if (idx == 1) {
+			up_write[0] = up;
+		} else if (up_vector_tilt_autofix_enabled && idx > 1) {
+			Vector3 normal = _find_convenient_axis_aligned_vec(up);
```

Comment:
Can you explain the purpose of this block in a high level language?

Reply:
the first block ensures that the first baked point will have the same up vector as the second one (if there's a second one)

the second block is the tilt autofix algorithm which progressively aligns the "sideways" axis (or x) with the xz plane of the "closest" orthogonal basis (actually, it's not the closest, I just copied the logic behind Basis' get_orthogonal_index to find a suitable orthogonal basis). It's merely aesthetic, the effects of it are shown in the gif on the main post.

Reply:
You essentially want to remove any accumulated holonomy, but this looks a bit hacky to me; you shouldn't be using orthogonal bases at all when doing that, those are for cubic structures

Reply:
My math knowledge is limited so I have no idea how else would I achieve this, it works just fine as it is, but if it seems too hacky I may remove it. It's not necessary after all. You can always adjust tilts by hand to achieve the same effect. Btw, the editor still has no means of doing it.

Reply:
@tagcup that's correct, but, unlike ROTATION_XY, it does in a progressive fashion, as if it was adjusting itself smoothly along the curve, not immediately. You can change how fast or how slow it will happen by changing the up_vector_tilt_autofix_speed property. 

also, I'm terrible at putting my thoughts into words, so I thank you for your patience ^^'

Reply:
ROTATION_XY should prevent all rotations happening along the curve-axis before they happen, so I'm not sure what you mean by "smoothly along the curve, not immediately.".

Reply:
Do you mean you want to accumulate some tilting smoothly and then also remove it smoothly?

If your purpose is to avoid tilting, what's the point of allowing it to happen in the first place?

Reply:
So I'd say that feature isn't necessary (if I'm not missing something here) and can be removed because ROTATION_XY should prevent tilting along the object-local forward/curve axis, assuming it's implemented correctly.

Reply:
>Do you mean you want to accumulate some tilting smoothly and then also remove it smoothly?

Yes, that's the point.

Reply:
Yeah, but what's the point of doing that?

Reply:
To be explicit, what is the point of accumulating some tilting around object-local Z smoothly, only to remove that very same tilting around object-local Z smoothly?

Reply:
Currently I only use this feature in loops, it makes it look naturap, at least from my perspective. Using the same algorithm to remove local z axis without making this smooth transition has noticeable side effects (the object suddenly tilts).

Reply:
> the object suddenly tilts

That shouldn't happen at intermediate times when following the path, assuming ROTATION_XY is working correctly.

Or do you mean at the beginning or at the end of the motion?

Reply:
Like, ROTATION_XY is supposed to lock all the rotations around Z, so if it tilts, it means the implementation is wrong

Reply:
You need to ensure you're removing the *object*-local Z rotations though, not parent-local or global. That's probably why it tilts

Reply:
I think it'd be good to explain from the beginning...

The current up vector generation algorithm has the following side effect when leaving a loop.

![no_autofix](https://user-images.githubusercontent.com/12879282/40079705-1c2eb8b0-585f-11e8-9b72-90c8575560f5.gif)

Noticed that the OrientedPathFollow is slightly tilted, and when it goes through other loops it tilts even more.

So I decided to create this autofix algorithm that would slightly tilt its local basis to the "closest" orthogonal basis out of the ones contained in matrix3.cpp. 

```
static const Basis _ortho_bases[24] = {
	Basis(1, 0, 0, 0, 1, 0, 0, 0, 1),
	Basis(0, -1, 0, 1, 0, 0, 0, 0, 1),
	Basis(-1, 0, 0, 0, -1, 0, 0, 0, 1),
	Basis(0, 1, 0, -1, 0, 0, 0, 0, 1),
	Basis(1, 0, 0, 0, 0, -1, 0, 1, 0),
	Basis(0, 0, 1, 1, 0, 0, 0, 1, 0),
	Basis(-1, 0, 0, 0, 0, 1, 0, 1, 0),
	Basis(0, 0, -1, -1, 0, 0, 0, 1, 0),
	Basis(1, 0, 0, 0, -1, 0, 0, 0, -1),
	Basis(0, 1, 0, 1, 0, 0, 0, 0, -1),
	Basis(-1, 0, 0, 0, 1, 0, 0, 0, -1),
	Basis(0, -1, 0, -1, 0, 0, 0, 0, -1),
	Basis(1, 0, 0, 0, 0, 1, 0, -1, 0),
	Basis(0, 0, -1, 1, 0, 0, 0, -1, 0),
	Basis(-1, 0, 0, 0, 0, -1, 0, -1, 0),
	Basis(0, 0, 1, -1, 0, 0, 0, -1, 0),
	Basis(0, 0, 1, 0, 1, 0, -1, 0, 0),
	Basis(0, -1, 0, 0, 0, 1, -1, 0, 0),
	Basis(0, 0, -1, 0, -1, 0, -1, 0, 0),
	Basis(0, 1, 0, 0, 0, -1, -1, 0, 0),
	Basis(0, 0, 1, 0, -1, 0, 1, 0, 0),
	Basis(0, 1, 0, 0, 0, 1, 1, 0, 0),
	Basis(0, 0, -1, 0, 1, 0, 1, 0, 0),
	Basis(0, -1, 0, 0, 0, -1, 1, 0, 0)
};
```

Since creating a basis for that, to check its orthogonal index would be an overkill, I decided to create this _find_convenient_axis_aligned_vec that mimicks its behavior. So I had this result.

![autofix](https://user-images.githubusercontent.com/12879282/40079698-169fbcbe-585f-11e8-84f7-8e769fcec113.gif)

This is the only thing that came in my mind to fix that "end of loop" tilt. If there's a better solution for that, I'd be glad to hear about.

Reply:
Oh, wait, the curve provides both forward and up, so the orientation of the object is fixed.

What you observe is totally normal and it's called holonomy.  It shouldn't be cancelled in such a manner; you should simply choose your up vector correctly if you want to end up with a particular orientation at the end. When you force it with a workaround like that, the object simply won't be following the provided curve, which beats the purpose of having the curve in the first place (note that with your modifications, curve not only defines a path but the full orientation).

Also, you can't implement ROTATION_Y and ROTATION_XY modes either without breaking the provided curve.

Overall, I'd say you should remove that feature because it's a hack, and it defeats the purpose of making the curve define the frame (because now it contains both forward and up), and bake your up vector carefully if you need to control the holonomy.

Reply:
And also remove the constraints because they break the curve frame

Reply:
As I said earlier, adjusting the tilt would be enough. So I'll remove both the autofix and the constraints, so... I think it's best to not inherit PathFollow, since I'd have to simply ignore its constraints property. What do you think?

Reply:
Yeah, that sounds sensible to me